File: Day085_2025/Q085_GFG.py
class Solution:
    def wordBreak(self, s: str, dictionary: list) -> bool:
        n = len(s)
        word_set = set(dictionary)  # Using a set for faster lookup
        dp = [False] * (n + 1)
        dp[0] = True

        for i in range(1, n + 1):
            for word in dictionary:
                start = i - len(word)
                if start >= 0 and dp[start] and s[start:i] == word:
                    dp[i] = True
                    break

        return dp[n]

File: Day085_2025/Q085_GFG.java
import java.util.*;

class Q085_GFG {
    public boolean wordBreak(String s, List<String> dictionary) {
        int n = s.length();
        // Set<String> wordSet = new HashSet<>(dictionary); // Using a HashSet for O(1) lookups
        boolean[] dp = new boolean[n + 1];
        dp[0] = true;

        for (int i = 1; i <= n; i++) {
            for (String word : dictionary) {
                int start = i - word.length();
                if (start >= 0 && dp[start] && s.substring(start, i).equals(word)) {
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[n];
    }
}
